core/src/main/java/org/apache/iceberg/DeleteFileIndex.java

```diff
@@ -582,93 +513,187 @@ private Iterable<CloseableIterable<ManifestEntry<DeleteFile>>> deleteManifestRea
     }
   }
 
-  // a group of indexed delete files sorted by the sequence number they apply to
-  private static class DeleteFileGroup {
-    private final long[] seqs;
-    private final IndexedDeleteFile[] files;
-
-    DeleteFileGroup(IndexedDeleteFile[] files) {
-      this.seqs = Arrays.stream(files).mapToLong(IndexedDeleteFile::applySequenceNumber).toArray();
-      this.files = files;
+  private static int findStartIndex(long[] seqs, long seq) {
+    int pos = Arrays.binarySearch(seqs, seq);
+    int start;
+    if (pos < 0) {
+      // the sequence number was not found, where it would be inserted is -(pos + 1)
+      start = -(pos + 1);
+    } else {
+      // the sequence number was found, but may not be the first
+      // find the first delete file with the given sequence number by decrementing the position
+      start = pos;
+      while (start > 0 && seqs[start - 1] >= seq) {
+        start -= 1;
+      }
     }
 
-    DeleteFileGroup(long[] seqs, IndexedDeleteFile[] files) {
-      this.seqs = seqs;
-      this.files = files;
+    return start;
+  }
+
+  private static DeleteFile[] concat(DeleteFile[]... deletes) {
+    return ArrayUtil.concat(DeleteFile.class, deletes);
+  }
+
+  // a group of position delete files sorted by the sequence number they apply to
+  private static class PositionDeletes {
+    private static final Comparator<DeleteFile> SEQ_COMPARATOR =
+        Comparator.comparingLong(DeleteFile::dataSequenceNumber);
+
+    private long[] seqs = null;
+    private DeleteFile[] files = null;
+    private volatile List<DeleteFile> buffer = Lists.newArrayList();
```

Comment:
This should be init'ed to null, to allow double check locking to work, right?

Reply:
Not really. We start by adding elements to the buffer list and then the group is indexed once needed.

Reply:
I see it now.  

One issue, it seems that if someone calls add() after some getter method, a NPE will be thrown?

Is there any real scenario where there are concurrent calls, or some instance where index is not needed?  I just looked briefly and thought referencedDeleteFiles() is called on every instance.  Just wondering if we can remove double-check locking, or even sort on insert instead of lazily, to make the class simpler?

Reply:
Yep, adding more elements upon indexing must throw an exception. I can add a precondition with a proper message but I was not sure it was worth it given that it will be called for each delete file and we may have millions of them. This is very internal code, not accessible from outside. I can add a precondition and benchmark it if you think that will be easier to follow.

I call `indexIfNeeded()` in all methods except `add()`, meaning all method triggered indexing. Let me know if you have other ideas, I would love to explore other options.

Reply:
Yea I was wondering if we could eliminate those member variables somehow.  I feel like we always need the sorted version in the end, so is there any need to be lazy?  
For example, always keep everything in a sorted treeset.  Or just have DeleteFileIndex::build() method populate intermediate  maps of Lists, and then have a final pass that makes EqualityDeletes/PositionDeletes with sorted arrays?

Yea if not possible to eliminate the member variables, i guess we can add the precondition and benchmark, hopefully it may get optimized out by branch prediciton

Reply:
We actually started with two passes but I migrated to one pass and indexing on demand for [these](https://github.com/apache/iceberg/pull/9251/files#r1426529884) reasons. Let me think a bit more on what can be done.

Reply:
I do like using `System.arraycopy` and binary search on arrays for lookups. I don't think we will beat that in terms of performance. The extra pass will also be expensive for path-based deletes and it will mean indexing of all deletes rather than indexing only what's covered by the query.

Let me start with a few comments and a precondition. I am going to think more about this.